### PR TITLE
Add support for state in HDF5

### PR DESCRIPTION
**Pull request by cjh1**
We convert the configuration to YAML and then use a custom loader/dumper to allow us to include numpy types in the HDF5 file in separate datasets. The high level YAML configuration is saved as a string dataset. We use the SafeLoader for loading so we have whitelisted a few safe python types.

This PR currently only persist the state held in `HexrdConfig`

Depends on: hexrd/hexrd#310.

### CODE REVIEW

File: hexrd/ui/state.py
Comment (cjh1) [inline hexrd/ui/state.py:34]: We may still want to validate the types here, to be extra safe ...

**Comment by psavery** [comment]
There are a lot of parts that seem to be working for saving/loading, namely:

1. Materials
2. Instrument
3. Hexrdgui config
4. Masks

We haven't tried data yet due to time constraints.

There may be some bugs that we encounter, as we have not had time to super thoroughly test this, but a lot of parts seem to be working for now, and we can perhaps review and merge it, after HEXRD/hexrd#310 gets merged.

https://user-images.githubusercontent.com/9558430/135114804-163b58b8-0c71-45cf-980b-b6b7541a93ee.mp4

**Comment by psavery** [comment]
This PR (along with HEXRD/hexrd#310) now supports saving/loading data to the state file as well!

Here is an example: [two_examples.zip](https://github.com/HEXRD/hexrdgui/files/7247959/two_examples.zip)

https://user-images.githubusercontent.com/9558430/135181682-dbfdef20-c303-475f-8ad8-2e95dedc2fe4.mp4

**Comment by joelvbernier** [comment]
Ok @cjh1 and @psavery, I found one bug so far; I loaded a materials file, an image set, and and instrument and set two overlays.  When I try to save, I get:
```
Traceback (most recent call last):
  File "/Users/bernier2/Documents/GitHub/hexrdgui/hexrd/ui/main_window.py", line 903, in on_action_save_state_triggered
    state.save(h5_file)
  File "/Users/bernier2/Documents/GitHub/hexrdgui/hexrd/ui/state.py", line 154, in save
    HexrdConfig().save_materials(h5_file, '/materials')
  File "/Users/bernier2/Documents/GitHub/hexrdgui/hexrd/ui/hexrd_config.py", line 682, in save_materials
    save_materials_hdf5(f, self.materials, path)
TypeError: save_materials_hdf5() takes 2 positional arguments but 3 were given
```

**Review by cjh1 — COMMENTED** [review]
@psavery This is looking good, I can approve it as I created the PR, if you approve we can get it merged.

**Review by psavery — APPROVED** [review]
LGTM

**Pull request merged**
